From b572a102dcce8154593fdf4af501ad4a7540d2c4 Mon Sep 17 00:00:00 2001
From: Maxime Leroy <maxime@leroys.fr>
Date: Fri, 5 Sep 2025 09:09:58 +0200
Subject: [PATCH 3/4] zebra: change params of get_iflink_speed

No needs to provide interface structure, only ifname and vrf_id are
useful. It will be used by a later commit.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>
---
 zebra/if_netlink.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/zebra/if_netlink.c b/zebra/if_netlink.c
index 1cfcc84..c8ee99f 100644
--- a/zebra/if_netlink.c
+++ b/zebra/if_netlink.c
@@ -256,13 +256,12 @@ static void netlink_vrf_change(struct nlmsghdr *h, struct rtattr *tb,
 		ctx, *(uint32_t *)RTA_DATA(attr[IFLA_VRF_TABLE]));
 }
 
-static uint32_t get_iflink_speed(struct interface *interface, int *error)
+static uint32_t get_iflink_speed(vrf_id_t vrf_id, const char *ifname, int *error)
 {
 	struct ifreq ifdata;
 	struct ethtool_cmd ecmd;
 	int sd;
 	int rc;
-	const char *ifname = interface->name;
 	uint32_t ret;
 
 	if (error)
@@ -281,7 +280,7 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 	/* use ioctl to get speed of an interface */
 	frr_with_privs(&zserv_privs) {
 		sd = vrf_socket(PF_INET, SOCK_DGRAM, IPPROTO_IP,
-				interface->vrf->vrf_id, NULL);
+				vrf_id, NULL);
 		if (sd < 0) {
 			if (IS_ZEBRA_DEBUG_KERNEL)
 				zlog_debug("Failure to read interface %s speed: %d %s",
@@ -292,7 +291,7 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 			return 0;
 		}
 		/* Get the current link state for the interface */
-		rc = vrf_ioctl(interface->vrf->vrf_id, sd, SIOCETHTOOL,
+		rc = vrf_ioctl(vrf_id, sd, SIOCETHTOOL,
 			       (char *)&ifdata);
 	}
 	if (rc < 0) {
@@ -320,7 +319,7 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 
 uint32_t kernel_get_speed(struct interface *ifp, int *error)
 {
-	return get_iflink_speed(ifp, error);
+	return get_iflink_speed(ifp->vrf->vrf_id, ifp->name, error);
 }
 
 static ssize_t
-- 
2.43.0

